src/modules/esl-popup/core/esl-popup.ts

```diff
@@ -74,88 +96,121 @@ export class ESLPopup extends ESLToggleable {
+  @bind
+  protected onActivatorIntersection(entries: IntersectionObserverEntry[], observer: IntersectionObserver) {
+    const entry = entries[0];
+    this._intersectionRatio = {};
+    if (!entry.isIntersecting) {
+      this.hide();
+      return;
+    }
 
-    // set arrow position
-    if (this.$arrow) {
-      this.$arrow.style.left = `${arrowLeft}px`;
-      this._arrowPosition = position;
+    if (entry.intersectionRect.y !== entry.boundingClientRect.y) {
+      this._intersectionRatio.top = entry.intersectionRect.height / entry.boundingClientRect.height;
+      this._checkIntersectionForAdjacentAxis(this._isPositioningAlongHorizontal, this._intersectionRatio.top);
+    }
+    if (entry.intersectionRect.bottom !== entry.boundingClientRect.bottom) {
+      this._intersectionRatio.bottom = entry.intersectionRect.height / entry.boundingClientRect.height;
+      this._checkIntersectionForAdjacentAxis(this._isPositioningAlongHorizontal, this._intersectionRatio.bottom);
+    }
+    if (entry.intersectionRect.x !== entry.boundingClientRect.x) {
+      this._intersectionRatio.left = entry.intersectionRect.width / entry.boundingClientRect.width;
+      this._checkIntersectionForAdjacentAxis(this._isPositioningAlongVertical, this._intersectionRatio.left);
+    }
+    if (entry.intersectionRect.right !== entry.boundingClientRect.right) {
+      this._intersectionRatio.right = entry.intersectionRect.width / entry.boundingClientRect.width;
+      this._checkIntersectionForAdjacentAxis(this._isPositioningAlongVertical, this._intersectionRatio.right);
     }
   }
 
-  protected _calculatePosition($activator: HTMLElement) {
-    const {left, arrowLeft} = this._calculateLeft($activator);
-    const {top, arrowTop, position} = this._calculateTop($activator);
-
-    return {
-      left,
-      top,
-      arrowLeft,
-      arrowTop,
-      position
-    };
+  @bind
+  protected onActivatorScroll(e: Event) {
+    this._updatePosition();
   }
 
-  protected _calculateLeft($activator: HTMLElement) {
-    const triggerRect = $activator.getBoundingClientRect();
-    const triggerPosX = triggerRect.left + window.pageXOffset;
-    const centerX = triggerPosX + triggerRect.width / 2;
+  protected _addActivatorObserver(target: HTMLElement) {
+    const scrollParents = getListScrollParents(target);
 
-    let arrowAdjust = 0;
-    let left = centerX - this.offsetWidth / 2;
+    const unsubscribers = scrollParents.map(($root) => {
+      const options = {passive: true} as EventListenerOptions;
+      $root.addEventListener('scroll', this.onActivatorScroll, options);
+      return () => {
+        $root && $root.removeEventListener('scroll', this.onActivatorScroll, options);
+      };
+    });
 
-    if (this.behavior === 'fit' && left < this._offsetWindow) {
-      arrowAdjust += left - this._offsetWindow;
-      left = this._offsetWindow;
-    }
+    const options = {
+      rootMargin: '0px',
+      threshold: [...Array(9).keys()].map((x) => x/8)
```

Comment:
Do we really use all of the breakpoints here? Big amount of breakpoints can affect performance.

Reply:
Quite a complicated question. Either we improve performance or we improve the smoothness and animations.
I had planned to take the issue of the behavior of the popup at the crossing points and leaving the window. But will make it a separate PR.
I guess need to experiment with rootMargin and thresholds to get a compromise result to satisfy performance and smoothness.